src/plugins/content_management/server/event_stream/es/init/es_event_stream_initializer.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import pRetry from 'p-retry';
+import { errors } from '@elastic/elasticsearch';
+import type { EsClient } from '../types';
+import type { EsEventStreamNames } from '../es_event_stream_names';
+import type { EventStreamLogger } from '../../types';
+import { newIndexTemplateRequest } from './index_template';
+
+export interface EsEventStreamInitializerDependencies {
+  names: EsEventStreamNames;
+  kibanaVersion: string;
+  logger: EventStreamLogger;
+  esClient: Promise<EsClient>;
+}
+
+export class EsEventStreamInitializer {
+  constructor(private readonly deps: EsEventStreamInitializerDependencies) {}
+
+  public async initialize(): Promise<void> {
+    const createdIndexTemplate = await this.#retry(
+      this.createIndexTemplateIfNotExists,
+      'createIndexTemplateIfNotExists'
+    );
+
+    if (createdIndexTemplate) {
+      await this.#retry(this.createDataStream, 'createDataStream');
+    }
+  }
+
+  /**
+   * Calls a function; retries calling it multiple times via p-retry, if it fails.
+   * Should retry on 2s, 4s, 8s, 16s.
+   *
+   * See: https://github.com/tim-kos/node-retry#retryoperationoptions
+   *
+   * @param fn Function to retry, if it fails.
+   */
+  readonly #retry = async <R>(fn: () => Promise<R>, fnName: string): Promise<R> => {
+    this.deps.logger.debug(`Event Stream initialization operation: ${fnName}`);
+
+    return await pRetry(fn, {
+      minTimeout: 1000,
+      maxTimeout: 1000 * 60 * 3,
+      retries: 4,
+      factor: 2,
+      randomize: true,
+      onFailedAttempt: (err) => {
+        const message =
+          `Event Stream initialization operation failed and will be retried: ${fnName};` +
+          `${err.retriesLeft} more times; error: ${err.message}`;
+
+        this.deps.logger.warn(message);
+      },
+    });
+  };
+
+  protected readonly createIndexTemplateIfNotExists = async (): Promise<boolean> => {
+    const exists = await this.indexTemplateExists();
+    if (exists) return false;
+    return await this.createIndexTemplate();
+  };
+
+  protected async indexTemplateExists(): Promise<boolean> {
+    try {
+      const esClient = await this.deps.esClient;
+      const name = this.deps.names.indexTemplate;
+      const exists = await esClient.indices.existsIndexTemplate({ name });
+
+      return !!exists;
```

Comment:
__nit:__ doesn't it return `boolean` already?
```suggestion
      return exists;
```

Reply:
It should be (but it is JavaScript, that's why only "should"). However, this cast will make it `boolean` for sure.